boa_engine/src/builtins/promise/mod.rs

```diff
@@ -0,0 +1,731 @@
+                    return Ok(JsValue::Undefined);
+                }
+
+                // 9. Let then be Completion(Get(resolution, "then")).
+                let then = resolution
+                    .as_object()
+                    .unwrap_or_else(|| unreachable!())
+                    .get("then", context);
+
+                let then = match then {
+                    // 10. If then is an abrupt completion, then
+                    Err(value) => {
+                        //   a. Perform RejectPromise(promise, then.[[Value]]).
+                        promise
+                            .borrow_mut()
+                            .as_promise_mut()
+                            .expect("Expected promise to be a Promise")
+                            .reject(&value, context)?;
+
+                        //   b. Return undefined.
+                        return Ok(JsValue::Undefined);
+                    }
+                    Ok(then) => then,
+                };
+
+                // 11. Let thenAction be then.[[Value]].
+                let then_action = then
+                    .as_object()
+                    .expect("rsolution.[[then]] should be an object")
+                    .get("Value", context)?;
+
+                // 12. If IsCallable(thenAction) is false, then
+                if !then_action.is_callable() {
+                    // a. Perform FulfillPromise(promise, resolution).
+                    promise
+                        .borrow_mut()
+                        .as_promise_mut()
+                        .expect("Expected promise to be a Promise")
+                        .fulfill(resolution, context)?;
+
+                    //   b. Return undefined.
+                    return Ok(JsValue::Undefined);
+                }
+
+                // 13. Let thenJobCallback be HostMakeJobCallback(thenAction).
+                let then_job_callback = JobCallback::make_job_callback(then_action);
+
+                // 14. Let job be NewPromiseResolveThenableJob(promise, resolution, thenJobCallback).
+                let job: JobCallback = PromiseJob::new_promise_resolve_thenable_job(
+                    promise.clone(),
+                    resolution.clone(),
+                    then_job_callback,
+                    context,
+                );
+
+                // 15. Perform HostEnqueuePromiseJob(job.[[Job]], job.[[Realm]]).
+                context.host_enqueue_promise_job(Box::new(job));
+
+                // 16. Return undefined.
+                Ok(JsValue::Undefined)
+            },
+            resolve_captures,
+        )
+        .name("")
+        .length(1)
+        .constructor(false)
+        .build();
+
+        // 5. Set resolve.[[Promise]] to promise.
+        resolve.set("Promise", promise.clone(), true, context)?;
+
+        // 6. Set resolve.[[AlreadyResolved]] to alreadyResolved.
+        resolve.set("AlreadyResolved", already_resolved.clone(), true, context)?;
```

Comment:
This initialization is unnecessary. The spec asks you to set `[[Promise]]` and `[[AlreadyResolved]]`, but those are the captures of `resolve`, which are initialized above. Also applies to the `reject` closure.

Reply:
This has been fixed in #2107.